internal/dig/shapes.lua

```diff
@@ -0,0 +1,289 @@
+-- shape definitions for gui/dig
+--@ module = true
+
+if not dfhack_flags.module then
+    qerror("this script cannot be called directly")
+end
+
+local to_pen = dfhack.pen.parse
+local CURSORS = {
+    INSIDE = { 1, 2 },
+    NORTH = { 1, 1 },
+    N_NUB = { 3, 2 },
+    S_NUB = { 4, 2 },
+    W_NUB = { 3, 1 },
+    E_NUB = { 5, 1 },
+    NE = { 2, 1 },
+    NW = { 0, 1 },
+    WEST = { 0, 2 },
+    EAST = { 2, 2 },
+    SW = { 0, 3 },
+    SOUTH = { 1, 3 },
+    SE = { 2, 3 },
+    VERT_NS = { 3, 3 },
+    VERT_EW = { 4, 1 },
+    POINT = { 4, 3 },
+}
+
+-- return the pen, alter based on if we want to display a corner and a mouse over corner
+function make_pen(direction, is_corner, is_mouse_over)
+    return to_pen {
+        ch = "X",
+        fg = COLOR_GREEN,
+        tile = dfhack.screen.findGraphicsTile(
```

Comment:
note that findGraphicsTile can be slow since it's a linear search across all tile pages. You might get better mileage out of pre-calculating all the texpos values and keeping them in a top-scope table. Note that this table should *not* be persisted across script invocations, though, since the texpos values can change if a world is unloaded and  a different world is loaded.

Reply:
Maybe that's why it was slow when I tried to change the render code to display the grid tiles on all non-shape tiles in the main script... if it's doing a linear search for every tile on the screen then that would make sense, I hadn't thought of caching the icons, is this also doing disk IO every time to load the icon from the image file too? Or are all the graphics loaded into memory on startup?

Reply:
no, it's all in memory. the linear search is for the "CURSORS" page, not the map tile (thankfully), but it's still linear.

In this case, I think the lag that you were seeing was due to the loop over all visible tiles. Lua just isn't as fast as C++ when doing that. I saw something similar in `gui/quickfort` when the blueprint preview covers a large area, and that doesn't do any `findGraphicsTile` lookups at all.

Reply:
That makes sense. Well, maybe if it's eventually possible to integrate with the vanilla dig layout then we get the grid overlay for free.

Reply:
> note that findGraphicsTile can be slow since it's a linear search across all tile pages. You might get better mileage out of pre-calculating all the texpos values and keeping them in a top-scope table. Note that this table should _not_ be persisted across script invocations, though, since the texpos values can change if a world is unloaded and a different world is loaded.

So if I'm understanding you right, I should do something like change the `CURSORS` table I'm creating in `shapes.lua` to contain the results of the calls to all the `findGraphicsTile` I expect to use. The second part I'm less sure about, would doing what I outlined avoid this issue? Or do I need to take extra steps to make sure it's not persisted. I don't know enough about the lifetime of the scripts between different states of the game.

Reply:
as a code example:
```
local function getPens()
    local pens = {}
    for _,n in ipairs{false, true} do for _,s in ipairs{false, true} do for _,e in ipairs{false, true} do for _,w in ipairs{false, true} do
        -- ...
    end end end end
    return pens
end
local PENS = getPens()
```
That will do the lookup once when the script runs when in a top-level script. It will *not* be re-run if in an internal module, so you'll have to calculate this in the top-level script and pass it to the functions called in this internal script

Reply:
So basically move all the pen/tile code to dig.lua instead of shapes.lua?